server/controllers/finance/reports/cashflow/index.js

```diff
@@ -753,3 +169,241 @@ function reportByService(req, res, next) {
+
+  _.extend(options, { orientation : 'landscape' });
+
+  // catch missing required parameters
+  if (!dateFrom || !dateTo || !cashboxesIds.length) {
+    throw new BadRequest('ERRORS.BAD_REQUEST');
+  }
+
+  try {
+    serviceReport = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    next(e);
+    return;
+  }
+
+  data.dateFrom = dateFrom;
+  data.dateTo = dateTo;
+
+  getCashboxesDetails(cashboxesIds)
+    .then(rows => {
+      data.cashboxes = rows;
+
+      data.cashAccountIds = data.cashboxes.map(cashbox => cashbox.account_id);
+
+      data.cashLabels = _.chain(data.cashboxes)
+        .map(cashbox => `${cashbox.label}`).uniq().join(' | ')
+        .value();
+
+      data.cashLabelSymbol = _.chain(data.cashboxes)
+        .map(cashbox => cashbox.symbol).uniq().join(' + ');
+
+      data.cashLabelDetails = data.cashboxes.map(cashbox => `${cashbox.account_number} - ${cashbox.account_label}`);
+
+      // get oposite lines to cashbox accounts in transactions as details
+      return getDetailsIdentifiers(data.cashAccountIds, dateFrom, dateTo);
+    })
+    .then(rows => {
+      data.detailsIdentifiers = rows.map(row => row.uuid);
+
+      // build periods columns from calculated period
+      return Fiscal.getPeriodsFromDateRange(data.dateFrom, data.dateTo);
+    })
+    .then(periods => {
+      data.periodDates = periods.map(p => p.start_date);
+
+      data.periods = periods.map(p => p.id);
+
+      data.colspan = data.periods.length + 1;
+
+      // skip period matrix if not transactions were identified
+      if (data.detailsIdentifiers.length === 0) { return []; }
+
+      // build periods string for query
+      const periodString = data.periods.length ? data.periods.map(periodId => {
+        return `SUM(CASE WHEN source.period_id = ${periodId} THEN source.balance ELSE 0 END) AS "${periodId}"`;
+      }).join(',') : '"NO_PERIOD" AS period';
+
+      const query = `
+        SELECT 
+          source.transaction_text, source.account_label, ${periodString},
+          source.transaction_type, source.transaction_id, source.account_id
+        FROM (
+          SELECT
+            a.number AS account_number, a.label AS account_label,
+            SUM(gl.debit_equiv - gl.credit_equiv) AS balance,
+            tt.id AS transaction_id, tt.type AS transaction_type,
+            MAX(tt.text) AS transaction_text,
+            gl.period_id, gl.account_id
+          FROM general_ledger gl
+          JOIN account a ON gl.account_id = a.id
+          JOIN transaction_type tt ON gl.origin_id = tt.id
+          WHERE gl.uuid IN ? 
+          GROUP BY a.id, tt.id
+        ) AS source 
+        GROUP BY transaction_type, account_id;
+      `;
+      return db.exec(query, [[data.detailsIdentifiers]]);
+    })
+    .then(rows => {
+      // split incomes from expenses
+      const incomes = _.chain(rows).filter({ transaction_type : 'income' }).groupBy('transaction_text').value();
```

Comment:
I haven't tried this yet, but will `_.partition` work ([docs](https://lodash.com/docs/#partition))?   It looks like you should be able to do something like this:
```js
const groups = _.partition(rows, row => transaction_type);
// this gives me an array of arrays [[...transactions with one type], [transactions with another type], ...]
// now we need to figure out what transaction type is in which array (they could be in any order).

// let's loop through the groups and convert them to a map of transaction type -> group
const mapping = {}
groups.forEach(group => {
   const [firstElementOfGroup] = group;
   mapping[firstElementOfGroup.transaction_type] = group;
});

// now we can do this:
const incomes = mapping.income.groupBy('transaction_text');
const expenses = mapping.expense.groupBy('transaction_text');
const transfers = mapping.transfer.groupBy('transaction_text');
```

It might also be more performant to use `groupBy('transaction_type_id')` instead of the `transaction_text`.  Comparing numbers is faster than comparing strings.

Reply:
Compare it to this : 
```js
const incomes = _.chain(rows).filter({ transaction_type : 'income' }).groupBy('transaction_text').value();
      const expenses = _.chain(rows).filter({ transaction_type : 'expense' }).groupBy('transaction_text').value();
      const transfers = _.chain(rows).filter({ transaction_type : 'transfer' }).groupBy('transaction_text').value();
```
It seems like there are a lot of code with your proposal.

Reply:
But much less loops for the poor Rsapberry Pi!